File: common/code/common/Http.cs
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;
using OpenTelemetry;
using OpenTelemetry.Metrics;
using OpenTelemetry.Trace;

namespace common;

public static class HttpModule
{
    public static void ConfigureTelemetry(OpenTelemetryBuilder builder) =>
        builder.WithTracing(tracing => tracing.AddHttpClientInstrumentation())
               .WithMetrics(metrics => metrics.AddHttpClientInstrumentation());

    public static void AddResilience(IHostApplicationBuilder builder) =>
        builder.Services.ConfigureHttpClientDefaults(builder => builder.AddStandardResilienceHandler());

    public static void AddServiceDiscovery(IHostApplicationBuilder builder) =>
        builder.Services
               .AddServiceDiscovery()
               .ConfigureHttpClientDefaults(builder => builder.AddServiceDiscovery());
}

File: common/code/common/Http.fs
[<RequireQualifiedAccess>]
module common.Http

open Microsoft.Extensions.Hosting
open Microsoft.Extensions.DependencyInjection
open global.OpenTelemetry
open OpenTelemetry.Trace
open OpenTelemetry.Metrics

let configureOpenTelemetryMetrics (metrics: MeterProviderBuilder) =
    metrics.AddHttpClientInstrumentation() |> ignore

let configureOpenTelemetryTracing (tracing: TracerProviderBuilder) =
    tracing.AddHttpClientInstrumentation() |> ignore

let addResilience (builder: IHostApplicationBuilder) =
    builder.Services.ConfigureHttpClientDefaults(fun builder -> builder.AddStandardResilienceHandler() |> ignore)
    |> ignore

let addServiceDiscovery (builder: IHostApplicationBuilder) =
    builder.Services
        .AddServiceDiscovery()
        .ConfigureHttpClientDefaults(fun builder -> builder.AddServiceDiscovery() |> ignore)
    |> ignore
